### ISSUE TITLE
[Enhancement] x360ce and DS4Windows-J2K Cooperation

### ISSUE THREAD

**Issue by SwinglesMan** [x360ce/x360ce#209]
Is there anyway that you can make it so that x360 cooperates alongside DS4-Windows-J2K? I would like to continue to use the features of the DS4Windows but I would also like to have the afterplugging/hotplugging ability that x360ce seems to now have.

Currently when using both, DS4Windows has priority over x350ce. Anyway to make it so that we can use certain things from both when they are both being used?

**Comment by Squall-Leonhart** [comment]
Use inputmapper, native xinput driver supports hotplugging.

**Comment by SwinglesMan** [comment]
Are you saying that Inputmapper makes games have hotplugging and afterplugging even if the game didn't have it at first (like x360ce)?

**Comment by Squall-Leonhart** [comment]
If the game doesn't have it, then theres no way x360ce can make it have it.

Hotplugging is not a feature of Directinput (and thus not a feature of the x360ce Xinput dll files), and is not possible unless using SDL layered over DInput, implementing  this would make x360ce too complex as then we'd be layering Xinput over SDL over Dinput.

**Comment by SwinglesMan** [comment]
The only reason I made this thread is that i see that it does give games this ability when they didn't have the ability at first. I tried x360ce with FFIII on steam, a game that doesn't have either, and with x360ce, it has these. If I remove x360ce, it doesn't have them. That's why I want it so that x360ce would take priority over ds4windows-j2k so that I can use the power saving features of ds4windows and the hotplugging/afterplugging features that x360ce now seem to have and be able to use them at the same time.

Edit: just tried it with dynasty warriors 8 on steam as well. Another game that doesn't have those features. And now they have them with x360ce being used but don't without x360ce.

**Comment by Squall-Leonhart** [comment]
Tapcio, did you implement hotplugging without mentioning?

**Comment by Nucleoprotein** [comment]
Nope, but hotplugging may work because x360ce does not detect if gamepad was disconnected and poll it anyway, so after reconnect it may work, ie. if this works it works because x360ce is too simple :p

**Comment by Squall-Leonhart** [comment]
I would think the device context would be lost normally.

I noticed FF14 has hot plugging detection, but it appears to be for DI, and doesn't work with X360CE Xinput.

However, the DS4 drivers are xinput native, so they should work with hotplugging when using wired.  Bluetooth is another issue all together.

**Comment by SwinglesMan** [comment]
Hey Tapeq, do you think you can modify a normal xinput.dll (original version without x360ce) to have it do what you just said and post it so I can test to see if it does the same that way too to test the hypothesis before doing anything to x360ce itself?

**Comment by Nucleoprotein** [comment]
Modify what ? Do what ? I see you have no idea how this stuff works ...
Short answer: impossible.

**Comment by SwinglesMan** [comment]
You said: "but hotplugging may work because x360ce does not detect if gamepad was disconnected and poll it anyway"

What I was saying was that if you could modify a normal xinput.dll to not detect if gamepad was disconnected and poll it anyway to test if we were on track.

I remember Krossx doing something similar by modifying a value in an xinput.dll to make it think a controller is always there and that made hotplugging work also (but not afterplugging though). 
He modified a value that always returns success, so there's no notification from the API that there's a controller plugged in. So if it checks for more controllers, all might show up as plugged in.

So that is why I had the idea that if you could make a normal version of the dll do what you said you think x360ce normally does, I could test it to see if the hypothesis was on point. 

@Squall-Leonhart : For me for FFIII, it works for bluetooth from my ds4 controller (bare controller, no ds4w) but not when I have it plugged in by cord. Weird.

**Comment by Nucleoprotein** [comment]
"but hotplugging may work because x360ce does not detect if gamepad was disconnected and poll it anyway" means it polls devices using DirectInput Poll/UpdateState, -but- it will return to game ERROR_DEVICE_NOT_CONNECTED like original xinput (https://github.com/x360ce/x360ce/blob/master/x360ce/x360ce/Controller.cpp#L77), so this is NOT designed to be hotpluging, **this is bug**, not a feature - as we want perfectly emulate/implement MS Xinput.

**Comment by SwinglesMan** [comment]
Ah, so that would explain why it only does that with the bare ds4 controller but not with an xinput program activated alongside it. Come to think of it, it was his modified Durazno, not the original dll (except it worked alongside xinput instead). 

How about if we keep this 'bug' but as a checkbox option used for games that lack hotplugging support for x360ce instead of getting rid of it completely? And expand upon it by doing similar 'bug' code on the xinput side so that it would be able to do this along side xinput programs like ds4w so that this hack option will separate the good code from the bug code and allow ppl to activate it on games that lack the feature?

**Comment by Squall-Leonhart** [comment]
Xinputs not ours, it belongs to microsoft XD

We just replicate it as well as we can for controllers that don't support it, if an application doesn't poll for state changes and survives a hotplug... well thats on the games end and we'll just let it happen i guess.

**Comment by EvoJocys** [comment]
This report is about the 3.x line, where x360ce was an xinput DLL copied into each game's folder. The current 4.x line works differently: it creates a virtual Xbox 360 controller that every game sees, with nothing copied into game folders. Game-specific problems from the DLL era do not carry over.

The latest release is 4.21 (September 2026): https://github.com/x360ce/x360ce/releases

Closing as part of a tracker clean-up after the 4.21 release. If it still happens on 4.21, reopen this issue with the version and what you see.
